R/devmode.R

```diff
@@ -0,0 +1,292 @@
+#'
+#' ```r
+#' # Reload the Shiny app when a sourced R file changes
+#' register_devmode_option(
+#'   "shiny.autoreload",
+#'   "Turning on shiny autoreload. To disable, call `options(shiny.autoreload = FALSE)`",
+#'   on = TRUE
+#' )
+#'
+#' # Use the unminified Shiny JavaScript file, `shiny.js`
+#' register_devmode_option(
+#'   "shiny.minified",
+#'   "Using full shiny javascript file. To use the minified version, call `options(shiny.minified = TRUE)`",
+#'   on = FALSE
+#' )
+#'
+#' # Display the full stack trace when errors occur during Shiny app execution
+#' register_devmode_option(
+#'   "shiny.fullstacktrace",
+#'   "Turning on full stack trace. To disable, call `options(shiny.fullstacktrace = FALSE)`",
+#'   on = TRUE
+#' )
+#' ```
+#'
+#' Other known, non-Shiny Developer Mode options:
+#'
+#' * Sass:
+#' ```r
+#' # Display the full stack trace when errors occur during Shiny app execution
+#' register_devmode_option(
+#'   "sass.cache",
+#'   "Turning off sass cache. To use default caching, call `options(sass.cache = TRUE)`",
+#'   on = FALSE
+#' )
+#' ```
+#' @param option Option to look for in `options()`
+#' @param message Message to display once every 8 hours if `option` is not set and if `in_devmode()` returns `TRUE` and Dev Mode messages can be displayed. For `get_devmode_option()`, if `message = waiver()`, the registered `message` value be attempted to be displayed.
+#' @param on Default value to return if `in_devmode()` returns `TRUE`. For `get_devmode_option()`, if `on = waiver()`, the registered default `on` value will be used.
+#' @examples
+#' # Ex: Within shiny, we register the option "shiny.minified"
+#' #   to default to `FALSE` when in Dev Mode
+#' \dontrun{register_devmode_option(
+#'   "shiny.minified",
+#'   message = paste0(
+#'     "Using full shiny javascript file. ",
+#'     "To use the minified version, call `options(shiny.minified = TRUE)`"
+#'   ),
+#'   on = FALSE
+#' )}
+#'
+register_devmode_option <- function(
+  option,
+  message = NULL,
+  on = NULL
+) {
+  if (!is.null(message)) {
+    stopifnot(length(message) == 1 && is.character(message))
+  }
+  registered_devmode_options$set(
+    option,
+    list(on = on, message = message)
+  )
+}
+registered_devmode_options
+
+#' @describeIn devmode Provides a consistent way to change default `getOption()` behavior when Developer Mode is enabled. This method is very similar to [getOption()] where the globally set option takes precedence. However, when the global option is not set, `default` will be returned when `in_devmode()` is `FALSE`, or `on` will be returned if `in_devmode()` is `TRUE`.
+#'
+#'
+#' @export
+#' @examples
+#' # Used within `shiny::runApp(launch.browser)`
+#' get_devmode_option("shiny.minified", TRUE) # TRUE if Dev mode is off
+#' is_minified <- with_devmode(TRUE, {
+#'   get_devmode_option("shiny.minified", TRUE)
+#' })
+#' is_minified # FALSE
+#'
+get_devmode_option <- function(
```

Comment:
Mke it clear that `devmode_default` and `devmode_message` generally should not be used; they should be set with `register_devmode_option()` (but doing it this way requires a Shiny dependency).